File: backend_py/sceneos_py/fal.py
"""
fal.ai LTX-Video provider — fast/cheap tier.

Mirrors backend/src/services/fal.ts. fal.subscribe blocks until the queue
finishes, so we wrap it in a fire-and-forget background task and surface
state through a local in-memory job map.
"""
from __future__ import annotations

import asyncio
import uuid
from typing import Any

from .config import env
from .provider import GenerateClipParams, StatusResult


_MODEL_ID = "fal-ai/ltx-video"
_JOBS: dict[str, dict[str, Any]] = {}
_configured = False


def _ensure_configured() -> None:
    global _configured
    if _configured:
        return
    key = env("FAL_API_KEY")
    if not key:
        raise RuntimeError(
            "fal.py: FAL_API_KEY is not set. Get a free key at https://fal.ai (signup includes credit)."
        )
    import os
    os.environ["FAL_KEY"] = key
    _configured = True


async def _run(provider_job_id: str, prompt: str, duration_seconds: float, image_url: str | None) -> None:
    try:
        import fal_client

        input_payload: dict[str, Any] = {
            "prompt": prompt,
            "duration_seconds": duration_seconds,
        }
        if image_url:
            input_payload["image_url"] = image_url

        # Run the blocking subscribe in a thread.
        result = await asyncio.to_thread(
            fal_client.subscribe,
            _MODEL_ID,
            arguments=input_payload,
            with_logs=False,
        )
        url = None
        if isinstance(result, dict):
            video = result.get("video")
            if isinstance(video, dict):
                url = video.get("url")
        if not url:
            _JOBS[provider_job_id] = {
                "status": "failed",
                "error": f"fal returned no video url. Raw: {str(result)[:200]}",
            }
            return
        _JOBS[provider_job_id] = {"status": "succeeded", "clipUrl": url}
    except Exception as exc:
        _JOBS[provider_job_id] = {"status": "failed", "error": str(exc)}


async def generate(params: GenerateClipParams) -> dict:
    _ensure_configured()
    provider_job_id = str(uuid.uuid4())
    _JOBS[provider_job_id] = {"status": "running"}

    asyncio.create_task(
        _run(
            provider_job_id,
            params["refinedPrompt"],
            params["durationSeconds"],
            params.get("startImageUrl"),
        )
    )
    return {"jobId": provider_job_id}


async def status(provider_job_id: str) -> StatusResult:
    job = _JOBS.get(provider_job_id)
    if not job:
        return {"status": "failed", "error": f"Unknown fal jobId: {provider_job_id}"}
    if job["status"] == "running":
        return {"status": "running"}
    if job["status"] == "failed":
        return {"status": "failed", "error": job.get("error", "fal failed")}
    return {"status": "succeeded", "clipUrl": job["clipUrl"]}

File: backend/src/services/fal.ts
/**
 * fal.ai video-generation service — "fal" tier.
 *
 * Tier: fast, real-AI, cheap. fal hosts LTX-Video (and many other models)
 * behind a single client. fal.subscribe() polls the queue server-side and
 * resolves once the job is complete — from our point of view that's a
 * single blocking promise rather than a real async lifecycle.
 *
 * To plug into the existing /api/generate → /api/status poll loop without
 * changing it, we wrap that promise into a fake job:
 *   1. generate() kicks off fal.subscribe() in the background, mints a
 *      UUID, stores the running state in JOBS keyed by that UUID, and
 *      returns the UUID immediately as the providerJobId.
 *   2. getStatus() looks up the UUID — still pending → "running",
 *      resolved → "succeeded" with the MP4 URL, rejected → "failed".
 *
 * The result MP4 is fal-hosted; the URL plays directly in browsers, which
 * is enough for the test-pipeline.sh end-to-end check. (Cloudinary upload
 * for downstream fl_splice happens later, when status.ts is wired to
 * thread clipPublicId through.)
 *
 * Reference:
 *   https://fal.ai/models/fal-ai/ltx-video
 *   https://docs.fal.ai/clients/javascript
 */
import { randomUUID } from "node:crypto";
import { fal } from "@fal-ai/client";
import type { JobStatus } from "../types/api.js";
import type { GenerateClipParams } from "./provider.js";

let configured = false;
function ensureConfigured() {
  if (configured) return;
  const key = process.env.FAL_API_KEY;
  if (!key) {
    throw new Error(
      "services/fal.ts: FAL_API_KEY is not set. Get a free key at https://fal.ai (signup includes $10 credit).",
    );
  }
  fal.config({ credentials: key });
  configured = true;
}

type FalJobState =
  | { status: "running" }
  | { status: "succeeded"; clipUrl: string }
  | { status: "failed"; error: string };

const JOBS = new Map<string, FalJobState>();

const MODEL_ID = "fal-ai/ltx-video";

interface LtxVideoInput {
  prompt: string;
  /** Optional first-frame conditioning. */
  image_url?: string;
  /** LTX accepts a target duration; clipped server-side to the model's range. */
  duration_seconds?: number;
}

interface LtxVideoOutput {
  video?: { url?: string };
}

export async function generate(
  params: GenerateClipParams & { startImageUrl?: string },
): Promise<{ jobId: string }> {
  ensureConfigured();

  const providerJobId = randomUUID();
  JOBS.set(providerJobId, { status: "running" });

  const input: LtxVideoInput = {
    prompt: params.refinedPrompt,
    duration_seconds: params.durationSeconds,
  };
  if (params.startImageUrl) input.image_url = params.startImageUrl;

  // Fire-and-forget: fal.subscribe() blocks until the queue completes, then
  // we flip the in-memory job state. The route returned the jobId already.
  void fal
    .subscribe(MODEL_ID, { input, logs: false })
    .then((result) => {
      const data = (result?.data ?? {}) as LtxVideoOutput;
      const url = data.video?.url;
      if (!url) {
        JOBS.set(providerJobId, {
          status: "failed",
          error: `fal returned no video url. Raw: ${JSON.stringify(result?.data ?? {}).slice(0, 200)}`,
        });
        return;
      }
      JOBS.set(providerJobId, { status: "succeeded", clipUrl: url });
    })
    .catch((err: unknown) => {
      const message = err instanceof Error ? err.message : String(err);
      JOBS.set(providerJobId, { status: "failed", error: message });
    });

  return { jobId: providerJobId };
}

export async function getStatus(
  providerJobId: string,
): Promise<{ status: JobStatus; clipUrl?: string; error?: string }> {
  const job = JOBS.get(providerJobId);
  if (!job) {
    return { status: "failed", error: `Unknown fal jobId: ${providerJobId}` };
  }
  if (job.status === "running") return { status: "running" };
  if (job.status === "failed") return { status: "failed", error: job.error };
  return { status: "succeeded", clipUrl: job.clipUrl };
}
